File: 3219-minimum-cost-for-cutting-cake-ii/3219-minimum-cost-for-cutting-cake-ii.py
class Solution:
    def minimumCost(self, m: int, n: int, horizontalCut: List[int], verticalCut: List[int]) -> int:
        horizontalCut.sort()
        verticalCut.sort()

        horizontalCut = horizontalCut[::-1]
        verticalCut   = verticalCut[::-1]
        
        h, v = 0, 0
        verticalPieces, horizontalPieces = 1, 1
        totalCost = 0
        
        while h < len(horizontalCut) and v < len(verticalCut):
            if horizontalCut[h] > verticalCut[v]:
                totalCost = totalCost + horizontalCut[h] * verticalPieces
                horizontalPieces += 1
                h += 1
            else:
                totalCost = totalCost + verticalCut[v] * horizontalPieces
                verticalPieces += 1
                v += 1
        
        while h < len(horizontalCut):
            totalCost = totalCost + horizontalCut[h] * verticalPieces
            h += 1
        
        while v < len(verticalCut):
            totalCost = totalCost + verticalCut[v] * horizontalPieces
            v += 1
        
        return totalCost

File: LeetCode/3219-minimum-cost-for-cutting-cake-ii/3219-minimum-cost-for-cutting-cake-ii.java
class Solution {
    public long minimumCost(int m, int n, int[] horizontalCut, int[] verticalCut) {
        Arrays.sort(horizontalCut);
        Arrays.sort(verticalCut);
        
        for (int i = 0; i < horizontalCut.length / 2; i++) {
            horizontalCut[i] = horizontalCut[i] ^ horizontalCut[horizontalCut.length - 1 - i];
            horizontalCut[horizontalCut.length - 1 - i] = horizontalCut[i] ^ horizontalCut[horizontalCut.length - 1 - i];
            horizontalCut[i] = horizontalCut[i] ^ horizontalCut[horizontalCut.length - 1 - i];
        }

        for (int i = 0; i < verticalCut.length / 2; i++) {
            verticalCut[i] = verticalCut[i] ^ verticalCut[verticalCut.length - 1 - i];
            verticalCut[verticalCut.length - 1 - i] = verticalCut[i] ^ verticalCut[verticalCut.length - 1 - i];
            verticalCut[i] = verticalCut[i] ^ verticalCut[verticalCut.length - 1 - i];
        }

        int h = 0, v = 0;
        long verticalPieces = 1, horizontalPieces = 1;
        long totalCost = 0;
        
        while (h < horizontalCut.length && v < verticalCut.length) {
            if (horizontalCut[h] > verticalCut[v]) {
                totalCost = totalCost + (long) horizontalCut[h] * verticalPieces;
                horizontalPieces++;
                h++;
            } else {
                totalCost = totalCost + (long) verticalCut[v] * horizontalPieces;
                verticalPieces++;
                v++;
            }
        }
        
        while (h < horizontalCut.length) {
            totalCost = totalCost + (long) horizontalCut[h] * verticalPieces;
            h++;
        }
        
        while (v < verticalCut.length) {
            totalCost = totalCost + (long) verticalCut[v] * horizontalPieces;
            v++;
        }
        
        return totalCost;
    }
}